designs/degraded-nodepools.md

```diff
@@ -0,0 +1,175 @@
+
+See below for example evaluations:
+
+```
+Successful Launch: 1
+Default: 0
+Unsuccessful Launch: -1
+
+[1, -1, 1, 1] = 3, `Degraded: False`
+[1, -1 , 1, -1] = 0, `Degraded: Unknown`
+[-1, -1, -1, 1] = -2, `Degraded: True`
+```
+
+One issue is determining when to clear or expire buffer entries. There are a few options:
+
+1) Clear the buffer after 3x the registration TTL to appropriately capture enough failures, even in small clusters
+2) Clear the buffer once every hour
+3) Expire one entry in the buffer after every evaluation but this approach would mean that zeroes are used to continue with the current status condition as opposed to going `Unknown`
+
+One major drawback of this approach is that it relies entirely on heuristics which can be wrong and mask failures.
+
+#### Considerations
+
+1. 👎👎 Due to heavy usage of heurstics, can mask Karpenter failures
+2. 👎 Over-engineered
+3. 👍 Keeps track of recent launches and addresses NodePool configurations that sometimes succeed
+4. 👍 Can be easily expanded to use enumerated types of failures
+
+### [Recommended] Option 4: Track if a NodePool/NodeClass Successfully Launched a Node
+
+This approach proposes tracking if a NodePool configuration has ever successfully launched using a `Verified` status.
+
+Given this would update only for the first node, this adds a neglible amount of apiserver traffic. There are a couple options to keep track of this info:
+
+1) Karpenter persists no info about past successes, only the current NodePool/NodeClass configuration. This likely exists on `NodePoolStatus` as `Verfied`. In this case, the status could be skipped and Karpenter can keep this in memory. If kept in memory, Karpenter would do the same thing as the initial addition of a NodePool and not write the status until it sees a node launch. If nodes are already provisioned using a NodePool then Karpenter can determine if there are nodes from the NodePool via the `Resources` status.
+
+```
+// NodePoolStatus defines the observed state of NodePool
+type NodePoolStatus struct {
+	// Resources is the list of resources that have been provisioned.
+	// +optional
+	Resources v1.ResourceList `json:"resources,omitempty"`
+	// LaunchedNode indicates if the current NodePool configuration has launched a node.
+	// +optional
+	Verified bool `json:"verified,omitempty"`
```

Comment:
```suggestion
	// Verified indicates if the current NodePool configuration has launched a node.
	// +optional
	Verified bool `json:"verified,omitempty"`
```

Using language like launch is interesting, the node has launched but to a user is it clear the differences between Launched, Registered, Initialized etc?  Are we going to use the Launched condition, or will we check that its gone through the full nodeclaim lifecycle before verifying a nodepool?  

Maybe they aren't configuring labels for daemonsets on their nodes correctly, so a daemonset dropping configuration for CNI never initializes and pods can't schedule, the nodes startup taints for that CNI are never are removed, but from the controllers perspective we have moved past a launch so its considered verified. 

Should we only consider a nodepool verified if a nodeclaim has gone through all the create lifecycle operations? (Launch + Initialization + Registration)

Reply:
>the node has launched but to a user is it clear the differences between Launched, Registered, Initialized etc?

>we have moved past a launch so its considered verified.

>Are we going to use the Launched condition, will we check that its gone through the full nodeclaim lifecycle before verifying a nodepool?

My mistake in wording - I think we consider a NodePool `Verified` after all NodeClaim lifecycle operations succeed for a given NodePool/NodeClass. 

As far as Karpenter could tell, everything _should_ have succeeded. Karpenter didn't see a cloudprovider create error but some other misconfiguration prevented a node from joining the cluster. If the NodePool/NodeClass has been verified then we can fire an event on the NodePool about the most recent failure with a probable cause. That could be the startup taints issue you mention or hitting registration TTLs because of no network path.

Reply:
Also, if a nodepool is updated then these verifications should be cleared.